inference-engine/src/vpu/common/src/ngraph/transformations/merge_gather_gather_elements.cpp

```diff
@@ -0,0 +1,145 @@
+// Copyright (C) 2021 Intel Corporation
+// SPDX-License-Identifier: Apache-2.0
+//
+
+#include "vpu/ngraph/transformations/merge_gather_gather_elements.hpp"
+
+#include <vpu/ngraph/operations/exp_gather_elements.hpp>
+#include <vpu/ngraph/utilities.hpp>
+
+#include <ngraph/pattern/op/wrap_type.hpp>
+#include <ngraph/opsets/opset6.hpp>
+#include <ngraph/pattern/matcher.hpp>
+#include <ngraph/pattern/op/label.hpp>
+#include <ngraph/validation_util.hpp>
+#include <ngraph/log.hpp>
+
+#include <numeric>
+
+NGRAPH_RTTI_DEFINITION(vpu::MergeGatherGatherElements, "MergeGatherGatherElements", 0);
+
+namespace vpu {
+
+bool MergeGatherGatherElements::run_on_function(std::shared_ptr<ngraph::Function> f) {
+    bool wasGraphChanged = false;
+
+    const auto gatherData = ngraph::pattern::any_input();
+    const auto gatherIndices = ngraph::pattern::any_input();
+    const auto gather = ngraph::pattern::wrap_type<ngraph::opset6::Gather>(
+        {gatherData, gatherIndices, ngraph::pattern::any_input()},
+        ngraph::pattern::consumers_count(1));
+
+    const auto squeezeAxes = ngraph::pattern::any_input(
+        ngraph::pattern::op::as_value_predicate(ngraph::pattern::has_class<ngraph::opset6::Constant>()));
+    const auto squeeze = ngraph::pattern::wrap_type<ngraph::opset6::Squeeze>({gather, squeezeAxes}, ngraph::pattern::consumers_count(1));
+
+    const auto transposePerm = ngraph::pattern::any_input(
+        ngraph::pattern::op::as_value_predicate(ngraph::pattern::has_class<ngraph::opset6::Constant>()));
+    const auto transpose = ngraph::pattern::wrap_type<ngraph::opset6::Transpose>({squeeze, transposePerm});
+
+    const auto m = std::make_shared<ngraph::pattern::Matcher>(transpose, "GatherSqueezeTransposeMatcher");
+    for (const auto& node : f->get_ordered_ops()) {
+        if (m->match(node)) {
+            auto& patternMap = m->get_pattern_value_map();
+
+            std::vector<ngraph::Node*> shapeOfs;
+            std::vector<ngraph::Node*> gatherElements;
+
+            const auto& m_transpose = patternMap.at(transpose);
+            const auto& transposeConsumers = m_transpose.get_target_inputs();
+            for (const auto& transposeConsumer : transposeConsumers) {
+                if (transposeConsumer.get_node()->get_type_info() == ngraph::opset6::ShapeOf::type_info) {
+                    shapeOfs.push_back(transposeConsumer.get_node());
+                } else if (transposeConsumer.get_node()->get_type_info() == ngraph::opset6::GatherElements::type_info) {
+                    gatherElements.push_back(transposeConsumer.get_node());
+                }
+            }
+            if (gatherElements.empty() || shapeOfs.size() + gatherElements.size() != transposeConsumers.size()) {
+                continue;
+            }
```

Comment:
Nitpick: you can use standard algorithm to check if there are any consumer which is neither ShapeOf nor GatherElements beforehand.
It may simplify code a bit

Reply:
Done.